File: backend/apps/core/permissions.py
"""
RBAC Permission classes for Vitali.
"""

from rest_framework.permissions import BasePermission

from apps.core.utils import tenant_has_feature


def is_platform_admin(user) -> bool:
    """
    True for Vitali platform operators — i.e. Django superusers.

    POLICY (operational, enforced outside the code): ``is_superuser`` is reserved
    for genuine Vitali platform staff. Tenant users — including clinic
    owners/admins — authorize exclusively via roles/permissions and must NEVER be
    created with ``is_superuser=True``. Under that policy, keying platform-admin
    powers off ``is_superuser`` is safe and keeps Django-admin semantics intact.

    The previous blanket bypass was scattered inline across permission classes;
    routing every check through this single helper makes the rule auditable and
    gives one place to harden later (e.g. a deploy-controlled allowlist or a
    dedicated flag) if defense-in-depth against a compromised superuser is needed.
    """
    return bool(
        user and getattr(user, "is_authenticated", False) and getattr(user, "is_superuser", False)
    )


class IsPlatformAdmin(BasePermission):
    """
    Grants access only to Vitali platform operators.

    Used for /api/v1/platform/* endpoints — plan management, subscriptions,
    module activation. Access is granted to Django superusers (Vitali platform
    operators); clinic owners authorize via roles and must never be superusers
    (see is_platform_admin for the policy).

    Usage:
        permission_classes = [IsAuthenticated, IsPlatformAdmin]
    """

    def has_permission(self, request, view):
        return is_platform_admin(request.user)


class ModuleRequiredPermission(BasePermission):
    """
    Checks that the current tenant has a specific module enabled via FeatureFlag.

    Vitali platform operators (see is_platform_admin) bypass module gating; a
    plain tenant superuser does not. Returns 403 with a clear message if the
    module is inactive.

    Usage:
        _BILLING = ModuleRequiredPermission('billing')
        permission_classes = [IsAuthenticated, _BILLING]

    Note: __call__ returns self so that DRF's get_permissions() works correctly
    when a pre-constructed instance is placed in permission_classes.
    """

    def __init__(self, module_key: str):
        self.module_key = module_key

    def __call__(self):
        return self

    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        if is_platform_admin(request.user):
            return True
        tenant = getattr(request, "tenant", None)
        if tenant is None:
            return False
        return tenant_has_feature(tenant, self.module_key)

    @property
    def message(self):
        return f"Module '{self.module_key}' is not active for this tenant."


class HasPermission(BasePermission):
    """
    Usage:
        permission_classes = [HasPermission('emr.read')]

    Reads the user's Role.permissions list and checks for the required perm.
    Role.permissions is stored as a JSON list: ["emr.read", "emr.write", ...]

    Note: __call__ returns self so that DRF's get_permissions() works correctly
    when a pre-constructed instance is placed in permission_classes.
    """

    def __init__(self, permission_required: str):
        self.permission_required = permission_required

    def __call__(self):
        return self

    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False
        if is_platform_admin(request.user):
            return True
        # Effective (per-tenant) role under Model B; falls back to the global role
        # when membership roles are not in effect. See User.effective_role.
        role = request.user.effective_role()
        if not role:
            return False
        return self.permission_required in role.permissions


# ─── Convenience factory ──────────────────────────────────────────────────────


def require_permission(perm: str):
    """
    Returns a HasPermission instance ready for permission_classes.
    Usage: permission_classes = [IsAuthenticated, require_permission('emr.read')]
    """
    return HasPermission(perm)


# ─── Default role permission sets ─────────────────────────────────────────────

ADMIN_PERMISSIONS = [
    "emr.read",
    "emr.write",
    "emr.sign",
    "emr.delete",
    "patients.read",
    "patients.write",
    "patients.delete",
    "billing.read",
    "billing.write",
    "billing.full",
    "schedule.read",
    "schedule.write",
    "pharmacy.read",
    "pharmacy.dispense",
    "pharmacy.full",
    "pharmacy.catalog_manage",
    "pharmacy.stock_manage",
    "pharmacy.warehouse_manage",
    "pharmacy.inventory_count",
    "pharmacy.inventory_approve",
    "pharmacy.transfer_manage",
    "pharmacy.transfer_accept",
    "pharmacy.recall_manage",
    "pharmacy.quarantine_manage",
    "pharmacy.dispense_controlled",
    "users.read",
    "users.write",
    "roles.read",
    "roles.write",
    "reports.read",
    "privacy.read",
    "privacy.manage",
    "workflow.read",
    "workflow.request",
    "workflow.approve",
    "ai.use",
    "ai.manage",
    "signatures.read",
    "signatures.sign",
    "fhir.read",
    "imaging.read",
    "imaging.write",
    "telemedicine.read",
    "telemedicine.host",
    "pharmacy_ai.read",
    "smart_scheduling.read",
    "triage.read",
    "triage.respond",
    "mobile.admin",
    "organization.read",
    "organization.write",
    "organization.delete",
    "mpi.read",
    "mpi.write",
    "mpi.review",
    "integrations.operations.read",
    "integrations.replay",
    "emar.read",
    "emar.administer",
    "sae.read",
    "sae.write",
    "pharmacy.clinical_validate",
]

CLINICAL_PRESCRIBER_PERMISSIONS = [
    "emr.read",
    "emr.write",
    "emr.sign",
    "patients.read",
    "patients.write",
    "billing.read",
    "schedule.read",
    "schedule.write",
    "pharmacy.read",
    "ai.use",
    "signatures.read",
    "signatures.sign",
    "fhir.read",
    "imaging.read",
    "imaging.write",
    "telemedicine.read",
    "telemedicine.host",
    "smart_scheduling.read",
    "triage.read",
    "emar.read",
    "sae.read",
]

NURSING_PERMISSIONS = [
    "emr.read",
    "emr.partial_write",
    "patients.read",
    "schedule.read",
    "pharmacy.dispense",
    "emar.read",
    "emar.administer",
    "sae.read",
    "sae.write",
]

RECEPTION_PERMISSIONS = [
    "patients.limited_read",
    "smart_scheduling.read",
    "schedule.read",
    "schedule.write",
    "billing.read",
    "triage.read",
    "triage.respond",
]

PHARMACY_PERMISSIONS = [
    "emr.read",
    "pharmacy.read",
    "pharmacy.dispense",
    "pharmacy.full",
    "pharmacy.catalog_manage",
    "pharmacy.stock_manage",
    "pharmacy.warehouse_manage",
    "pharmacy.inventory_count",
    "pharmacy.inventory_approve",
    "pharmacy.transfer_manage",
    "pharmacy.transfer_accept",
    "pharmacy.recall_manage",
    "pharmacy.quarantine_manage",
    "pharmacy.dispense_controlled",
    "patients.limited_read",
    "pharmacy_ai.read",
    "pharmacy.clinical_validate",
]

BILLING_PERMISSIONS = [
    "billing.read",
    "billing.write",
    "billing.full",
    "patients.limited_read",
    "emr.read",
    "ai.use",
]

DEFAULT_ROLES = {
    "admin": ADMIN_PERMISSIONS,
    "medico": CLINICAL_PRESCRIBER_PERMISSIONS,
    "enfermeiro": NURSING_PERMISSIONS,
    "recepcao": RECEPTION_PERMISSIONS,
    "recepcionista": RECEPTION_PERMISSIONS,
    "farmaceutico": PHARMACY_PERMISSIONS,
    "faturista": BILLING_PERMISSIONS,
    "dentista": CLINICAL_PRESCRIBER_PERMISSIONS,
}

File: frontend/lib/permissions.ts
/**
 * Canonical RBAC permission catalog for the Vitali frontend.
 *
 * SINGLE SOURCE OF TRUTH for nav↔permission mapping. Every string here mirrors a
 * permission the backend actually enforces — do NOT invent strings the backend
 * does not check. Derived from `backend/apps/core/permissions.py`:
 *   - `ADMIN_PERMISSIONS` (the exhaustive admin role set),
 *   - the per-role sets (`CLINICAL_PRESCRIBER_PERMISSIONS`, `NURSING_PERMISSIONS`,
 *     `RECEPTION_PERMISSIONS`, `PHARMACY_PERMISSIONS`, `BILLING_PERMISSIONS`) —
 *     which add `patients.limited_read`, `emr.partial_write`,
 *   - and literals passed to `HasPermission("…")` across the apps
 *     (`hr.manage` is enforced by `apps/hr/views.HRAccessPermission` but is not in
 *     the admin list, so it is included here explicitly).
 *
 * The menu is UX only (defense in depth). The real security boundary stays the
 * backend `permission_classes`; hiding a menu item is never an access control.
 */

import type { UserDTO } from "@/lib/auth";

export const PERMISSIONS = {
  // Tenant-admin capability (non-forgeable; see role_has_admin_capability)
  ADMIN: "admin",

  // EMR / clinical
  EMR_READ: "emr.read",
  EMR_WRITE: "emr.write",
  EMR_PARTIAL_WRITE: "emr.partial_write",
  EMR_SIGN: "emr.sign",
  EMR_DELETE: "emr.delete",

  // Patients
  PATIENTS_READ: "patients.read",
  PATIENTS_LIMITED_READ: "patients.limited_read",
  PATIENTS_WRITE: "patients.write",
  PATIENTS_DELETE: "patients.delete",

  // Scheduling
  SCHEDULE_READ: "schedule.read",
  SCHEDULE_WRITE: "schedule.write",
  SMART_SCHEDULING_READ: "smart_scheduling.read",

  // Billing / financial
  BILLING_READ: "billing.read",
  BILLING_WRITE: "billing.write",
  BILLING_FULL: "billing.full",

  // Pharmacy / supplies
  PHARMACY_READ: "pharmacy.read",
  PHARMACY_DISPENSE: "pharmacy.dispense",
  PHARMACY_DISPENSE_CONTROLLED: "pharmacy.dispense_controlled",
  PHARMACY_FULL: "pharmacy.full",
  PHARMACY_CATALOG_MANAGE: "pharmacy.catalog_manage",
  PHARMACY_STOCK_MANAGE: "pharmacy.stock_manage",
  PHARMACY_WAREHOUSE_MANAGE: "pharmacy.warehouse_manage",
  PHARMACY_INVENTORY_COUNT: "pharmacy.inventory_count",
  PHARMACY_INVENTORY_APPROVE: "pharmacy.inventory_approve",
  PHARMACY_TRANSFER_MANAGE: "pharmacy.transfer_manage",
  PHARMACY_TRANSFER_ACCEPT: "pharmacy.transfer_accept",
  PHARMACY_RECALL_MANAGE: "pharmacy.recall_manage",
  PHARMACY_QUARANTINE_MANAGE: "pharmacy.quarantine_manage",
  PHARMACY_CLINICAL_VALIDATE: "pharmacy.clinical_validate",
  PHARMACY_AI_READ: "pharmacy_ai.read",

  // Medication administration / nursing
  EMAR_READ: "emar.read",
  EMAR_ADMINISTER: "emar.administer",
  SAE_READ: "sae.read",
  SAE_WRITE: "sae.write",

  // Imaging / diagnostics
  IMAGING_READ: "imaging.read",
  IMAGING_WRITE: "imaging.write",
  FHIR_READ: "fhir.read",
  TRIAGE_READ: "triage.read",
  TRIAGE_RESPOND: "triage.respond",
  TELEMEDICINE_READ: "telemedicine.read",
  TELEMEDICINE_HOST: "telemedicine.host",

  // People & operations (HR)
  HR_MANAGE: "hr.manage",

  // Users / roles administration
  USERS_READ: "users.read",
  USERS_WRITE: "users.write",
  ROLES_READ: "roles.read",
  ROLES_WRITE: "roles.write",

  // Reporting / workflow / privacy
  REPORTS_READ: "reports.read",
  WORKFLOW_READ: "workflow.read",
  WORKFLOW_REQUEST: "workflow.request",
  WORKFLOW_APPROVE: "workflow.approve",
  PRIVACY_READ: "privacy.read",
  PRIVACY_MANAGE: "privacy.manage",

  // Organization / identity (MPI)
  ORGANIZATION_READ: "organization.read",
  ORGANIZATION_WRITE: "organization.write",
  ORGANIZATION_DELETE: "organization.delete",
  MPI_READ: "mpi.read",
  MPI_WRITE: "mpi.write",
  MPI_REVIEW: "mpi.review",

  // Integrations / signatures / AI / mobile
  INTEGRATIONS_OPERATIONS_READ: "integrations.operations.read",
  INTEGRATIONS_REPLAY: "integrations.replay",
  SIGNATURES_READ: "signatures.read",
  SIGNATURES_SIGN: "signatures.sign",
  AI_USE: "ai.use",
  AI_MANAGE: "ai.manage",
  MOBILE_ADMIN: "mobile.admin",
} as const;

/** A backend-enforced permission string. */
export type Permission = (typeof PERMISSIONS)[keyof typeof PERMISSIONS];

/** Flat list of every catalog permission (deduped). */
export const ALL_PERMISSIONS: Permission[] = Array.from(
  new Set(Object.values(PERMISSIONS)),
) as Permission[];

/**
 * Visibility gate carried by a nav entry. All three dimensions are optional and
 * evaluated in order by {@link canSee}. Absent dimensions do not restrict.
 */
export interface NavGate {
  /** Module/tier key checked against the tenant's active modules (FeatureFlag). */
  module?: string;
  /** RBAC permissions — visible if the user holds AT LEAST ONE (`.some`). */
  permissions?: Permission[];
  /** Platform (superuser) gate — only for Vitali platform operators. */
  superuser?: boolean;
}

/**
 * Decide whether a nav entry is visible for `user`, given the tenant's resolved
 * `activeModules` (null = still loading).
 *
 * Gate order and fail-safe policy (see UI_NAVIGATION_IA.md §6):
 *   1. Module/tier — fail-OPEN while modules load (`null`) to avoid flicker;
 *      once resolved, a missing module hides the item.
 *   2. Superuser — fail-SAFE: hidden unless `is_superuser === true` (an absent
 *      signal keeps Plataforma hidden by default).
 *   3. RBAC permission — NEVER fail-open: a user without at least one listed
 *      permission is hidden. This is the dominant gate.
 *
 * This is UX only; the backend `permission_classes` remains the real barrier.
 */
export function canSee(
  user: Pick<UserDTO, "permissions" | "is_superuser">,
  gate: NavGate,
  activeModules: string[] | null,
): boolean {
  // Gate 1 — module/tier (fail-open while loading).
  if (gate.module && activeModules !== null && !activeModules.includes(gate.module)) {
    return false;
  }

  // Gate 2 — superuser / Plataforma (fail-safe: undefined → hidden).
  if (gate.superuser && user.is_superuser !== true) {
    return false;
  }

  // Gate 3 — RBAC permission (never fail-open).
  if (gate.permissions && gate.permissions.length > 0) {
    const held = user.permissions ?? [];
    if (!gate.permissions.some((permission) => held.includes(permission))) {
      return false;
    }
  }

  return true;
}
